### PR TITLE
feat(organizations): create endpoints to handle organization invitations TASK-969

### PR DESCRIPTION
**Pull request by rajpatel24**
### 🗒️ Checklist

1. [x] run linter locally
2. [x] update all related docs (API, README, inline, etc.), if any
3. [x] draft PR with a title `<type>(<scope>)<!>: <title> TASK-1234`
4. [x] tag PR: at least `frontend` or `backend` unless it's global
5. [x] fill in the template below and delete template comments
6. [x] review thyself: read the diff and repro the preview as written
7. [x] open PR & confirm that CI passes
8. [ ] request reviewers, if needed
9. [ ] delete this section before merging

### 📣 Summary
Implemented endpoints for organization invitations, allowing organization owners to invite existing users or unregistered users to join their organization. The invitee can either accept or decline the invitation. If the invitee accepts, their assets will be transferred to the organization.

### 📖 Description

- Organization owners can send invitations to users (both registered and unregistered) via email or username.
- The invitee can accept or decline the invitation. If accepted, the invitee's assets will be transferred to the organization owner.

> POST https://[kpi]/api/v2/organizations/org_12345/invites/

- Create organization invites for registered and unregistered users.
- Set the role for which the user is being invited - (Choices: `member`, `admin`). Default is `member`.

Payload:
```
{
    "invitees": ["demo14", "demo13@demo13.com", "demo25@demo25.com"],
    "role": "admin"
}
```

Response:
```
[
    {
        "url": "http://kf.kobo.local/api/v2/organizations/org_12345/invites/7a4f9a3b-9112-43cc-a6ae-bb4a6583b4b2/",
        "invited_by": "http://kf.kobo.local/api/v2/users/gtl_raj/",
        "status": "pending",
        "invitee_role": "admin",
        "created": "2025-01-06T13:01:40Z",
        "modified": "2025-01-06T13:01:40Z",
        "invitee": "demo14"
    },
    {
        "url": "http://kf.kobo.local/api/v2/organizations/org_12345/invites/6746121a-7a87-4c2d-9994-85e38d8cff65/",
        "invited_by": "http://kf.kobo.local/api/v2/users/gtl_raj/",
        "status": "pending",
        "invitee_role": "admin",
        "created": "2025-01-06T13:01:40Z",
        "modified": "2025-01-06T13:01:40Z",
        "invitee": "demo13"
    },
    {
        "url": "http://kf.kobo.local/api/v2/organizations/org_12345/invites/2af706a9-4f67-4145-a0d3-8d66fbc77a19/",
        "invited_by": "http://kf.kobo.local/api/v2/users/gtl_raj/",
        "status": "pending",
        "invitee_role": "admin",
        "created": "2025-01-06T13:01:40Z",
        "modified": "2025-01-06T13:01:40Z",
        "invitee": "demo25@demo25.com"
    }
]
```

> GET https://[kpi]/api/v2/organizations/org_12345/invites/

Response:
```
{
    "count": 3,
    "next": null,
    "previous": null,
    "results": [
        {
            "url": "http://kf.kobo.local/api/v2/organizations/org_12345/invites/6746121a-7a87-4c2d-9994-85e38d8cff65/",
            "invited_by": "http://kf.kobo.local/api/v2/users/gtl_raj/",
            "status": "pending",
            "invitee_role": "admin",
            "created": "2025-01-06T13:01:40Z",
            "modified": "2025-01-06T13:01:40Z",
            "invitee": "demo13"
        },
        {
            "url": "http://kf.kobo.local/api/v2/organizations/org_12345/invites/7a4f9a3b-9112-43cc-a6ae-bb4a6583b4b2/",
            "invited_by": "http://kf.kobo.local/api/v2/users/gtl_raj/",
            "status": "pending",
            "invitee_role": "admin",
            "created": "2025-01-06T13:01:40Z",
            "modified": "2025-01-06T13:01:40Z",
            "invitee": "demo14"
        },
        {
            "url": "http://kf.kobo.local/api/v2/organizations/org_12345/invites/2af706a9-4f67-4145-a0d3-8d66fbc77a19/",
            "invited_by": "http://kf.kobo.local/api/v2/users/gtl_raj/",
            "status": "pending",
            "invitee_role": "admin",
            "created": "2025-01-06T13:01:40Z",
            "modified": "2025-01-06T13:01:40Z",
            "invitee": "demo25@demo25.com"
        }
    ]
}
```

> PATCH https://[kpi]/api/v2/organizations/org_12345/invites/7a4f9a3b-9112-43cc-a6ae-bb4a6583b4b2/

- Update an organization invite to accept, decline, cancel, expire, or resend.

Payload:
```
{
    "status": "accepted"
}
```

Response:
```
{
    "url": "http://kf.kobo.local/api/v2/organizations/org_12345/invites/7a4f9a3b-9112-43cc-a6ae-bb4a6583b4b2/",
    "invited_by": "http://kf.kobo.local/api/v2/users/gtl_raj/",
    "status": "accepted",
    "invitee_role": "admin",
    "created": "2025-01-06T13:01:40Z",
    "modified": "2025-01-06T13:01:40Z",
    "invitee": "demo14"
}
```

> DELETE https://[kpi]/api/v2/organizations/org_12345/invites/7a4f9a3b-9112-43cc-a6ae-bb4a6583b4b2/

- Organization owner or admin can delete an organization invite.

Response: 204

### CODE REVIEW

**Comment by magicznyleszek** [comment]
@rajpatel24 Wuld that whole invite object be present in `/api/v2/organizations/:organization_id/members/` in the `invite` property?

File: kobo/apps/organizations/models.py [hunk 1]
```diff
@@ -46,6 +47,19 @@ class OrganizationType(models.TextChoices):
...
+
+    ACCEPTED = 'accepted'
+    CANCELLED = 'cancelled'
+    COMPLETE = 'complete'
+    DECLINED = 'declined'
+    EXPIRED = 'expired'
+    FAILED = 'failed'
+    IN_PROGRESS = 'in_progress'
+    PENDING = 'pending'
+    RESENT = 'resent'
```
Comment (magicznyleszek) [inline kobo/apps/organizations/models.py:60]: @rajpatel24 What's the difference between `in_progress` and `pending`?

**Comment by rajpatel24** [comment]
> @rajpatel24 Wuld that whole invite object be present in `/api/v2/organizations/:organization_id/members/` in the `invite` property? And I see you have `pending` status in there, and is this the same as `sent`?

@magicznyleszek No, the invite object would not be present there. I don't think it's necessary since `/api/v2/organizations/:organization_id/members/` is designed to list the members who are already part of the organization - those who have accepted the invitations and are now part of it.

The `pending` status is the same as `sent`. When the organization owner sends an invitation, the status of the invite will be set to `pending`.

File: kobo/apps/organizations/models.py (hunk 1 above)
Comment (rajpatel24) [reply kobo/apps/organizations/models.py:60]: @magicznyleszek As of now, we're not using `in_progress` as a status for organization invitations. I included it as a possible choice, along with `completed` and `failed`, in case the need arises in the future. Once @noliveleger  is back (probably today), I'll discuss this with him. Currently, when the organization owner sends an invitation, the status is set to `pending`. When the invitee accepts the invitation, the status changes to `accepted` in the `OrganizationInvitation` table.

Additionally, when a user accepts an invitation, we trigger an async task to transfer their projects to the organization. This is part of the existing functionality we're reusing for project transfers along with organization invitations. It uses its own `project-ownership/Invite` table with different statuses, which you might already be familiar with.

So, my question for @noliveleger  would be: when the project transfer is completed and the status in the `project-ownership/Invite` table is set to `completed`, do we also need to set the status to `completed` in the `OrganizationInvitation` table, or is the `accepted` state sufficient?

**To summarize**, the `/api/v2/organizations/:organization_id/invites/` API will currently have the following statuses: `pending`, `accepted`, `declined`, `cancelled`, `expired`, and `resent`.

If any changes occur regarding this, I'll make sure to update you right away.

Comment (noliveleger) [reply kobo/apps/organizations/models.py:61]: @rajpatel24 , `accepted` would be enough for `OrganizationInvitation`.

File: kobo/apps/organizations/templates/emails/accepted_invite.txt [hunk 2]
```diff
@@ -0,0 +1,10 @@
+{% load i18n %}
+{% trans "Dear" %} {{ sender_username }},
+
+{% blocktrans %}{{ recipient_username }} ({{ recipient_email }}) has accepted your request to join {{ organization_name }} organization.{% endblocktrans %}
+
+{% trans "All projects, submissions, data storage, transcription and translation usage for their projects will be transferred to you." %}
+
+{% trans "Note: You will continue to have permissions to manage these projects until the user permissions are changed." %}
```
Comment (noliveleger) [inline kobo/apps/organizations/templates/emails/accepted_invite.txt:8]: I don't think this sentence should be here since that's the sender who receives this message.
(Please fix HTML template too)

File: kobo/apps/organizations/templates/emails/registered_user_invite.txt [hunk 3]
```diff
@@ -0,0 +1,18 @@
+{% load i18n %}
+{% trans "Hello," %}
```
Comment (noliveleger) [inline kobo/apps/organizations/templates/emails/registered_user_invite.txt:2]: Let's be a little more fancy,
If email has been sent to a username, let's use: `Hello <username>`
if email has been sent to a valid email with only one account, let's use: `Hello <username>`  
if email has been sent to a valid email with more than one account, let's use: `Hello <email>`  .

Then get rid of `(username: {{ recipient_username }}) ` in block below


What happens if invitation is sent to an email with multiple account? I don't see the template for that? The text is little bit different

**Review by noliveleger — CHANGES_REQUESTED** [review]

File: kobo/apps/organizations/templates/emails/registered_user_invite.txt (hunk 3 above)
Comment (rajpatel24) [reply kobo/apps/organizations/templates/emails/registered_user_invite.txt]: Sure, I've updated the template accordingly. 

Regarding the template for multiple accounts, it is included in the same template with the `has_multiple_accounts` condition. I’ve also updated some text for it.

File: kobo/apps/organizations/views.py [hunk 4]
```diff
@@ -419,4 +461,195 @@ def get_queryset(self):
...
+    http_method_names = ['get', 'post', 'patch', 'delete']
+    lookup_field = 'guid'
+
+    def get_queryset(self):
+        organization_id = self.kwargs['organization_id']
+        return OrganizationInvitation.objects.select_related(
+            'invitee', 'invited_by', 'organization'
+        ).filter(organization_id=organization_id)
+
+    def create(self, request, *args, **kwargs):
```
Comment (noliveleger) [inline kobo/apps/organizations/views.py:646]: a->z :-)

**Review by noliveleger — CHANGES_REQUESTED** [review]

File: kobo/apps/organizations/views.py [hunk 5]
```diff
@@ -419,4 +461,195 @@ def get_queryset(self):
...
+    >
+    >       curl -X DELETE https://[kpi]/api/v2/organizations/org_12345/invites/f3ba00b2-372b-4283-9d57-adbe7d5b1bf1/
+
+    > Response 204
+
+    """
+    serializer_class = OrgMembershipInviteSerializer
+    permission_classes = [OrgMembershipInvitePermission]
+    http_method_names = ['get', 'post', 'patch', 'delete']
+    lookup_field = 'guid'
```
[the remark this answers is not in the document]
Comment (rajpatel24) [reply kobo/apps/organizations/views.py:639]: @noliveleger I tried to override the `guid` field from `django-organizations` with our `KpiUidField`, but the parent class’s `save()` method includes logic to set a unique ID if `guid` is absent. This prevents `KpiUidField` from saving the intended unique ID.

To address this, we might need to bypass the parent class’s save() method by doing:
```
def save(self, **kwargs):
    models.Model.save(self, **kwargs)
```

However, I’m concerned about potential trade-offs. Another option would be to remove the `guid` field and add a completely new `uid` field. What do you suggest?

[the remark this answers is not in the document]
Comment (noliveleger) [reply kobo/apps/organizations/views.py:639]: let's keep `guid` as-is. Forget about our `uid`.

File: kobo/apps/organizations/models.py [hunk 6]
```diff
@@ -365,21 +376,26 @@ def send_invite_email(self):
         email_message = EmailMessage(
             to=to_email,
             subject=t(
-                f"You're invited to join {organization_name}'s organization"
-            ),
+                f"You're invited to join ##organization_name## organization"
```
Comment (noliveleger) [inline kobo/apps/organizations/models.py:379]: Does not need to be an f-string anymore

File: kobo/apps/organizations/models.py [hunk 7]
```diff
@@ -365,21 +376,26 @@ def send_invite_email(self):
...
 
     def send_refusal_email(self):
+        """
+        Send an email to the sender of the invitation to notify them that the
+        invitee has declined the invitation
+        """
+        sender_language = self.invited_by.extra_details.data.get(
+            'last_ui_language', 'en'
+        )
+        activate(sender_language)
```
Comment (noliveleger) [inline kobo/apps/organizations/models.py:397]: It seems it would be better to use 

```python
with translation.override(sender_language):
    ...
```

instead of `activate(sender_language)`

Otherwise, everytime  we use `activate()` we should restore the current language when we are done with translations


```python
current_language = django.utils.translation.get_language()
activate(sender_language)
# do what you want with translations
activate(current_language)
```

File: kobo/apps/organizations/models.py [hunk 8]
```diff
@@ -365,21 +376,26 @@ def send_invite_email(self):
...
         Mailer.send(email_message)
 
     def send_refusal_email(self):
+        """
+        Send an email to the sender of the invitation to notify them that the
+        invitee has declined the invitation
+        """
+        sender_language = self.invited_by.extra_details.data.get(
+            'last_ui_language', 'en'
+        )
```
Comment (noliveleger) [inline kobo/apps/organizations/models.py:396]: instead of hard-coding `'en'`, we should use either `settings.LANGUAGE_CODE` or `django.utils.translation.get_language()`


```python 
current_language = django.utils.translation.get_language()
# or current_language = settings.LANGUAGE_CODE
sender_language = self.invited_by.extra_details.data.get(
            'last_ui_language', current_language
)
```

File: kobo/apps/organizations/models.py [hunk 9]
```diff
@@ -365,21 +376,26 @@ def send_invite_email(self):
         email_message = EmailMessage(
             to=to_email,
             subject=t(
-                f"You're invited to join {organization_name}'s organization"
-            ),
+                f"You're invited to join ##organization_name## organization"
+            ).replace('##organization_name##', organization_name),
```
Comment (noliveleger) [inline kobo/apps/organizations/models.py:380]: You could use your utility `replace_placeholders()`

File: kobo/apps/organizations/serializers.py [hunk 10]
```diff
@@ -338,6 +340,7 @@ def _send_status_email(self, instance, status):
         if email_func:
             email_func()
 
+    @void_cache_for_request(keys=("organization",))
```
Comment (noliveleger) [inline kobo/apps/organizations/serializers.py:343]: I'm surprised the linter did not catch it but use single quotes 🙏 .

File: kobo/apps/organizations/tests/test_organization_invitations.py [hunk 11]
```diff
@@ -63,57 +65,86 @@ def _update_invite(self, user, guid, status):
...
+        if response.status_code == status.HTTP_200_OK:
+            self.assertEqual(response.data['status'], 'resent')
+            self.assertEqual(mail.outbox[0].to[0], invitation.invitee.email)
 
-    def test_owner_can_cancel_invitation(self):
+    @data(
+        ('owner', status.HTTP_200_OK),
+        ('admin', status.HTTP_200_OK),
+        ('member', status.HTTP_403_FORBIDDEN),
+        ('external', status.HTTP_403_FORBIDDEN)
```
Comment (noliveleger) [inline kobo/apps/organizations/tests/test_organization_invitations.py:129]: I think it should 404 for external. We don't want to reveal the existence of the organization to external.

File: kobo/apps/organizations/tests/test_organization_invitations.py [hunk 12]
```diff
@@ -63,57 +65,86 @@ def _update_invite(self, user, guid, status):
...
+                    mail.outbox[index].to[0],
+                    invite.email if invite else invitation['invitee']
+                )
 
-    def test_owner_can_resend_invitation(self):
+    @data(
+        ('owner', status.HTTP_200_OK),
+        ('admin', status.HTTP_200_OK),
+        ('member', status.HTTP_403_FORBIDDEN),
+        ('external', status.HTTP_403_FORBIDDEN)
```
Comment (noliveleger) [inline kobo/apps/organizations/tests/test_organization_invitations.py:104]: I think it should 404 for external. We don't want to reveal the existence of the organization to external.

**Review by noliveleger — CHANGES_REQUESTED** [review]

File: kobo/apps/organizations/models.py (hunk 9 above)
Comment (rajpatel24) [reply kobo/apps/organizations/models.py]: Done. Also I have moved the utility function to a new file, as it is a generic utility not specific to the organizations app.
